### DISCUSSION TITLE
Compute metrics every n batches

### DISCUSSION THREAD

**Discussion by Ch-rode** [Lightning-AI/pytorch-lightning#12487] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
Hello ! I'm trying to use Pytorch lightning but I don't have clear in my mind all the steps. Anyway I'm trying to calculate the train_loss (for example) not only for each step(=batch) or epochs but **every n batches** (i.e. 500) but I'm not sure how to compute it (compute, reset etc). I tried this approach but this is not working. Can you help me? thanks 

Any other suggestion on the code is well appreciated !

```
def training_step(self, batch: tuple, batch_nb: int, *args, **kwargs) -> dict:
        """ 
        Runs one training step. This usually consists in the forward function followed
            by the loss function.
        
        :param batch: The output of your dataloader. 
        :param batch_nb: Integer displaying which batch this is
        Returns:
            - dictionary containing the loss and the metrics to be added to the lightning logger.
        """
        inputs, targets = batch
        model_out = self.forward(**inputs)
        loss_val = self.loss(model_out, targets)
        


        y = targets["labels"]
        y_hat = model_out["logits"]
        
        labels_hat = torch.argmax(y_hat, dim=1)
        val_acc = self.metric_acc(labels_hat, y)
        tqdm_dict = {"train_loss": loss_val, 'batch_nb': batch_nb}
        self.log('train_loss', loss_val, on_step=True, on_epoch=True,prog_bar=True)
        self.log('train_acc', val_acc, on_step=True, prog_bar=True,on_epoch=True)

        # reset the metric to restart accumulating
        self.loss_val_bn = self.loss(model_out, targets) #accumulate state
        if batch_nb % 500 == 0:
            self.log("x batches test loss_train", self.loss_val_bn.compute(),batch_nb) # perform a compute every 10 batches
            self.loss_val_bn.reset() 


        #output = OrderedDict(
            #{"loss": loss_val, "progress_bar": tqdm_dict, "log": tqdm_dict})

        # can also return just a scalar instead of a dict (return loss_val)
        #return output
        return loss_val
```

**Comment by rohitgr7** [comment]
hey @Ch-rode !

what is `self.loss` here?
asking because you are calling it twice here:
```py
loss_val = self.loss(model_out, targets)
...
self.loss_val_bn = self.loss(model_out, targets) #accumulate state
```
is it a torchmetric instance?

**Comment by Ch-rode** [comment]
Thank you for your answer. 
yes i'm calling double only to check if with .compute and .reset the result will be different.
The loss is defined as following:
```

def __build_loss(self):
        """ Initializes the loss function/s. """
        self._loss = nn.CrossEntropyLoss()

def loss(self, predictions: dict, targets: dict) -> torch.tensor:
        """
        Computes Loss value according to a loss function.
        :param predictions: model specific output. Must contain a key 'logits' with
            a tensor [batch_size x 1] with model predictions
        :param labels: Label values [batch_size]
        Returns:
            torch.tensor with loss value.
        """
        return self._loss(predictions["logits"], targets["labels"])`
```

**Reply by rohitgr7** [reply]
```py
return self._loss(predictions["logits"], targets["labels"])`
```
returns tensor. I don't think it will have `.compute` and `.reset` since that is available with torchmetrics.

so you want to accumulate the loss for 500 steps, log it and reset it and do this again and again? or just log it at every 500 steps?

**Reply by Ch-rode** [reply]
My idea is to have during the training not only the metrics for each step (so for each batch if i'm correct) or each epoch in the end, but also a way to check if my training loss is increasing or decreasing as the progress with the training. Indeed it just came to my mind that also reset is not a good idea because will compute the loss only for the subsample of batches.
Tell me if iI have not expressed myself clearly! As I understood log_every_n_steps will print only the loss for the specified step not an overall trend.

**Reply by rohitgr7** [reply]
`log_every_n_steps` will somewhat do the same. you need to log at every step, but `log_every_n_steps` takes care of when to actually log it to the logger. So if you set `log_every_n_steps=500`, it will log the loss at every 500 steps, not just at the 500th step. So as the training progresses it will display the trend for you.

**Reply by Ch-rode** [reply]
Ah ok sorry I think I misunderstood the documentation, so it will log main the aggregate loss/acc following the trend until I reach the end of the epoch, am I right?
```
 def training_step(self, batch: tuple, batch_nb: int, *args, **kwargs) -> dict:

        inputs, targets = batch
        model_out = self.forward(**inputs)
        loss_val = self.loss(model_out, targets)
        
        y = targets["labels"]
        y_hat = model_out["logits"]
        
        labels_hat = torch.argmax(y_hat, dim=1)
        val_acc = self.metric_acc(labels_hat, y)
        tqdm_dict = {"train_loss": loss_val, 'batch_nb': batch_nb}
        self.log('train_loss', loss_val, on_step=True, on_epoch=True,prog_bar=True)
        self.log('train_acc', val_acc, on_step=True, prog_bar=True,on_epoch=True)

        #output = OrderedDict(
            #{"loss": loss_val, "progress_bar": tqdm_dict, "log": tqdm_dict})

        # can also return just a scalar instead of a dict (return loss_val)
        #return output
        return loss_val
```

Using:
```
trainer = Trainer(
    log_every_n_steps=1000,
    gpus=hparams.gpus,
    tpu_cores=hparams.tpu_cores,
    logger=logger,
    callbacks=early_stop_callback,
    strategy="dp",
    max_epochs=hparams.max_epochs,
    min_epochs=hparams.min_epochs,
    accumulate_grad_batches=hparams.accumulate_grad_batches,
    limit_val_batches=hparams.val_percent_check,
    checkpoint_callback = checkpoint_callback,
    precision=hparams.precision,
    amp_level=hparams.amp_level,
    deterministic=True,
    amp_backend="apex",
    auto_lr_find=False,     
    auto_scale_batch_size=False,
    progress_bar_refresh_rate=50
)
```

**Reply by rohitgr7** [reply]
> so it will log main the aggregate loss/acc

not the aggregated loss, but the loss computed at that step.

**Reply by Ch-rode** [reply]
OK thanks but this is indeed what I don't want. I want information about my loss during the training, how is its trend, so how is increasing or decreasing overall taking into account also the previous batches. (my batch is 1 for example). Like an average loss until that moment.

**Reply by rohitgr7** [reply]
then you can probably do something like:
```py
class AverageMeter(object):
    def __init__(self):
        self.reset()

    def reset(self):
        self.val = 0
        self.count = 0

    def update(self, val):
        self.val += val
        self.count += 1

    def compute(self):
        return self.val / self.count
```
and
```py
class LitModel(LightningModule):
    def __init__(self, ...):
        super().__init__()
        self.loss_meter = AverageMeter()

    def training_step(self, ...):
        loss = ...
        self.loss_meter.update(loss)
        
        if self.global_step % 500 == 0:
            self.log('avg_loss', self.loss_meter.compute())
            self.loss_meter.reset()
```
you can also update the `AverageMeter` to compute the weighted average if you want that.
I could have recommended using torchmetrics, but losses are not available there.

**Reply by Ch-rode** [reply]
Thank you for your help. Can I also ask you how can I log/print this and in general information on my metrics.csv? I'm using TestTube 
thank you ! 

```
def setup_testube_logger() -> TestTubeLogger:
    """ Function that sets the TestTubeLogger to be used. """
    now = datetime.now()
    dt_string = now.strftime("%d-%m-%Y--%H-%M-%S")

    return TestTubeLogger(
        save_dir="experiments/",
        version=dt_string,
        name="lightning_logs",
    )

logger = setup_testube_logger()
```

**Reply by rohitgr7** [reply]
it will be logged to the csv when this is called:
```py
self.log('avg_loss', self.loss_meter.compute())
```

**Reply by Ch-rode** [reply]
I tried but actually it is not printed nothing after each 500 batches.

**Reply by rohitgr7** [reply]
what's the value for log_every_n_steps in your case?

**Reply by Ch-rode** [reply]
This is my metric.csv after Epoch1:
![Schermata da 2022-03-29 15-01-06](https://user-images.githubusercontent.com/61243245/160617068-8adffa5f-9c8d-43b7-8659-f8532fe37c6c.png)
And this is my code for the logging part
```
    def training_step(self, batch: tuple, batch_nb: int, *args, **kwargs) -> dict:

        inputs, targets = batch
        model_out = self.forward(**inputs)
        loss_val = self.loss(model_out, targets)

        self.loss_meter.update(loss_val)
        if self.global_step % 1000 == 0:
            self.log('loss_avg_over_batches', self.loss_meter.compute())
            self.log("batch_nb",batch_nb)
            #self.loss_meter.reset()
        

        y = targets["labels"]
        y_hat = model_out["logits"]
        
        labels_hat = torch.argmax(y_hat, dim=1)
        val_acc = self.metric_acc(labels_hat, y)
        self.log('train_loss', loss_val, on_step=True, on_epoch=True,prog_bar=True,logger=True)
        self.log('train_acc', val_acc, on_step=True, prog_bar=True,on_epoch=True,logger=True)

        # can also return just a scalar instead of a dict (return loss_val)
        #return output
        return loss_val


    def validation_step(self, batch: tuple, batch_nb: int, *args, **kwargs) -> dict:

        inputs, targets = batch

        model_out = self.forward(**inputs)
        loss_val = self.loss(model_out, targets)


        y = targets["labels"]
        y_hat = model_out["logits"]
        
        labels_hat = torch.argmax(y_hat, dim=1)
        val_acc = self.metric_acc(labels_hat, y)

        #tqdm_dict = {"val_loss": loss_val, 'val_accuracy': val_acc, 'batch_nb': batch_nb}
        #self.log("val_step",tqdm_dict, on_epoch=True) 


        self.log('val_loss', loss_val) #pytorch bug
        self.log('val_acc', val_acc)


        output = OrderedDict({"val_loss": loss_val, "val_acc": val_acc})

        return output

    
    def validation_epoch_end(self, outputs: list) -> dict:

        val_loss_mean = torch.stack([x['val_loss'] for x in outputs]).mean()
        val_acc_mean = torch.stack([x['val_acc'] for x in outputs]).mean()

        tqdm_dict = {"avg_val_acc": val_acc_mean,  'avg_val_loss': val_loss_mean, 'epoch': self.current_epoch}

        self.log("log",tqdm_dict, on_epoch=True, prog_bar=True,logger=True) 

        result = {
            "progress_bar": tqdm_dict,
            "log": tqdm_dict,
            "val_loss": val_loss_mean,
        }
        return result
......

trainer = Trainer(
    log_every_n_steps=1000,
......

**Reply by Ch-rode** [reply]
and the progress_bar is something like this, a complete mess ! Thank you very much for your time

```
Epoch 1:  25%|██▌       | 8700/34484 [1:27:39<4:19:47,  1.65it/s, loss=0.656, v_num=5-11, train_loss_step=1.180, train_acc_step=0.000, log={'avg_val_acc': 0.6954756379127502, 'avg_val_loss': 0.5897443294525146, 'epoch': 0.0}, train_loss_epoch=0.638, train_acc_epoch=0.664]
```

**Reply by rohitgr7** [reply]
I tried and it seems to be working for me.

Also, I tried [CSVLogger](https://pytorch-lightning.readthedocs.io/en/latest/extensions/generated/pytorch_lightning.loggers.CSVLogger.html#csvlogger) since TestTubeLogger is deprecated.

![Screenshot 2022-03-30 at 11 49 43 PM](https://user-images.githubusercontent.com/30778939/160904294-79cef7d3-64da-4c4c-b269-6b8faae882b3.png)

> and the progress_bar is something like this, a complete mess !

because this is not the right way:
```py
self.log("log",tqdm_dict, on_epoch=True, prog_bar=True,logger=True) 
```
instead try:
```py
self.log_dict(tqdm_dict, on_epoch=True, prog_bar=True,logger=True) 
```
ref: https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#log-dict

**Reply by Ch-rode** [reply]
Can I kindly ask you to share your code to have such output (for the training_step/valifation_step/validation_epoch_end)? I tried what you did but for sure i'm doing something wrong because nothuings changes, espcially I have only 1 row for epoch 0. MY pytorch lightning version is 1.6.0rc0

**Reply by rohitgr7** [reply]
here, it's very basic:
```py
import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.loggers import CSVLogger


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class AverageMeter(object):
    def __init__(self):
        self.reset()

    def reset(self):
        self.val = 0
        self.count = 0

    def update(self, val):
        self.val += val
        self.count += 1

    def compute(self):
        return self.val / self.count


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)
        self.loss_meter = AverageMeter()

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        self.loss_meter.update(loss)
        if self.global_step % 1000 == 0:
            self.log('loss_avg_over_batches', self.loss_meter.compute())
            self.loss_meter.reset()
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 6400), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 6400), batch_size=2)
    logger = CSVLogger(save_dir='.')

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        logger=logger,
        num_sanity_val_steps=0,
        max_epochs=5,
        enable_model_summary=False,
        log_every_n_steps=1000,
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)


if __name__ == "__main__":
    run()
```

**Reply by Ch-rode** [reply]
Indeed, I also re-edit again my code as yours but still not printing nothing after 1000 steps. I also removed the validation_epoch_end as I read on the forum because now it is not needed anymore if you use self.log.
```
def training_step(self, batch: tuple, batch_nb: int, *args, **kwargs) -> dict:

        inputs, targets = batch
        model_out = self.forward(**inputs)
        loss_val = self.loss(model_out, targets)
        self.log("train_loss", loss_val)

        self.loss_meter.update(loss_val)
        if self.global_step % 1000 == 0:
            self.log('loss_avg_over_batches', self.loss_meter.compute())
        
        return {"loss": loss_val}

def validation_step(self, batch: tuple, batch_nb: int, *args, **kwargs) -> dict:

        inputs, targets = batch

        model_out = self.forward(**inputs)
        loss_val = self.loss(model_out, targets)


        y = targets["labels"]
        y_hat = model_out["logits"]
        
        labels_hat = torch.argmax(y_hat, dim=1)
        val_acc = self.metric_acc(labels_hat, y)

        self.log("valid_loss", loss_val)
        self.log("valid_acc", val_acc)
        

trainer = Trainer(
    log_every_n_steps=1000,
    gpus=hparams.gpus,
    tpu_cores=hparams.tpu_cores,
    logger=logger,
    callbacks=early_stop_callback,
    strategy="dp",
    max_epochs=hparams.max_epochs,
    min_epochs=hparams.min_epochs,
    accumulate_grad_batches=hparams.accumulate_grad_batches,
    limit_val_batches=hparams.val_percent_check,
    checkpoint_callback = checkpoint_callback,
    precision=hparams.precision,
    amp_level=hparams.amp_level,
    deterministic=True,
    amp_backend="apex",
    auto_lr_find=False,     
    auto_scale_batch_size=False,
    enable_model_summary=False,
    #progress_bar_refresh_rate=50
)
```
Really sorry to bother you but I'm trying to understand !

**Reply by rohitgr7** [reply]
can you try running my code shared above and see if it works for you?
if it does, then I'd probably need a small reproducible example to check the issue on your side. It's not that hard :)

**Reply by Ch-rode** [reply]
I just tried your code as it is and it works perfectly, as you shared metrics.csv with me 
i do not understand which can be the problem
